In [ ]:
import pandas as pd
import numpy as np
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

In [ ]:
from statsforecast.models import AutoARIMA
from statsforecast import StatsForecast

d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [3]:
df = pd.read_csv("D:\\Users\\Md Mahfuzur Rahman\\Desktop\\Research\\for_me\\cleaned_combined_data.csv", parse_dates=["datetime"])
df["unique_id"] = "BD"
df.head()

,datetime,demand_mw,load_shedding,temp_mean,wspd_mean,unique_id
0,2020-01-01 03:00:00,5351,0,18.65,1.80,BD
1,2020-01-01 06:00:00,5346,0,23.73,3.00,BD
2,2020-01-01 09:00:00,6654,0,25.25,4.58,BD
3,2020-01-01 12:00:00,6750,0,21.05,2.17,BD
4,2020-01-01 15:00:00,6546,0,18.82,0.30,BD


In [4]:
df['hour'] = df['datetime'].dt.hour
df['dayofweek'] = df['datetime'].dt.dayofweek
df['month'] = df['datetime'].dt.month
df['dayofyear'] = df['datetime'].dt.dayofyear

# -----------------------------
# Cyclic time features
# -----------------------------

df['hour_sin'] = np.sin(2 * np.pi * df['hour'] / 24)
df['hour_cos'] = np.cos(2 * np.pi * df['hour'] / 24)

df['dow_sin'] = np.sin(2 * np.pi * df['dayofweek'] / 7)
df['dow_cos'] = np.cos(2 * np.pi * df['dayofweek'] / 7)

df['month_sin'] = np.sin(2 * np.pi * df['month'] / 12)
df['month_cos'] = np.cos(2 * np.pi * df['month'] / 12)

df['doy_sin'] = np.sin(2 * np.pi * df['dayofyear'] / 365)
df['doy_cos'] = np.cos(2 * np.pi * df['dayofyear'] / 365)


# -----------------------------
# Lag features
# -----------------------------

for lag in [1, 2, 3, 6, 12, 24, 48, 72, 168, 336]:
    df[f'lag{lag}'] = df['demand_mw'].shift(lag)


# -----------------------------
# Rolling features
# IMPORTANT: shift(1)
# prevents target leakage
# -----------------------------

past_demand = df['demand_mw'].shift(1)

for window in [6, 12, 24, 48, 168]:

    df[f'roll{window}_mean'] = (
        past_demand.rolling(window).mean()
    )

    df[f'roll{window}_std'] = (
        past_demand.rolling(window).std()
    )


# Remove NaN
df = df.dropna().reset_index(drop=True)

In [5]:
df_ext = (
    df[["datetime", "demand_mw",

        'hour_sin','hour_cos','dow_sin','dow_cos','month_sin','month_cos','doy_sin','doy_cos',
    
        'load_shedding','temp_mean','wspd_mean',
    
        'lag1','lag2','lag3','lag6','lag12','lag24','lag48','lag72','lag168','lag336',
    
        'roll6_mean','roll6_std','roll12_mean','roll12_std','roll24_mean','roll24_std',
        'roll48_mean','roll48_std','roll168_mean','roll168_std']]
    .rename(columns={"datetime": "ds", "demand_mw": "y"})
    .assign(unique_id=df["unique_id"])
)

In [11]:
def metrics(name, y_true, y_pred):
    y_true = np.asarray(y_true, dtype=float).ravel()
    y_pred = np.asarray(y_pred, dtype=float).ravel()

    mae  = mean_absolute_error(y_true, y_pred)
    rmse = np.sqrt(mean_squared_error(y_true, y_pred))
    mape = np.mean(np.abs((y_true - y_pred) / y_true)) * 100
    r2   = r2_score(y_true, y_pred)
    acc  = max(0.0, 1.0 - mape / 100.0) * 100
    return dict(Model=name,
                Accuracy=round(acc, 3),
                MAE=round(mae, 1),
                RMSE=round(rmse, 1),
                MAPE=round(mape, 3),
                R2=round(r2, 4))

results = []  # Store results for all models

In [8]:
# Sort chronologically
sorted_df = df_ext.sort_values("ds").reset_index(drop=True)

# Last 168 rows for testing
test_size = 168

test = sorted_df.iloc[-test_size:].copy()

# Remaining data
remaining_df = sorted_df.iloc[:-test_size].copy()

# Last 20% of the remaining data for training
train_size = int(len(remaining_df) * 0.20)

train = remaining_df.iloc[-train_size:].copy()

print("Train shape:", train.shape)
print("Test shape:", test.shape)

print("Train time range:",
      train["ds"].min(), "to", train["ds"].max())

print("Test time range:",
      test["ds"].min(), "to", test["ds"].max())

print("\nTrain tail:")
print(train.tail())

print("\nTest:")
print(test)

Train shape: (8421, 34)
Test shape: (168, 34)
Train time range: 2024-12-18 08:00:00 to 2025-12-24 16:00:00
Test time range: 2025-12-24 17:00:00 to 2025-12-31 23:00:00

Train tail:
                       ds      y      hour_sin  hour_cos   dow_sin   dow_cos  \
42102 2025-12-24 12:00:00  10134  1.224647e-16 -1.000000  0.974928 -0.222521   
42103 2025-12-24 13:00:00  10545 -2.588190e-01 -0.965926  0.974928 -0.222521   
42104 2025-12-24 14:00:00  10209 -5.000000e-01 -0.866025  0.974928 -0.222521   
42105 2025-12-24 15:00:00   9967 -7.071068e-01 -0.707107  0.974928 -0.222521   
42106 2025-12-24 16:00:00   9629 -8.660254e-01 -0.500000  0.974928 -0.222521   

          month_sin  month_cos   doy_sin   doy_cos  ...   roll6_std  \
42102 -2.449294e-16        1.0 -0.120208  0.992749  ...  748.203560   
42103 -2.449294e-16        1.0 -0.120208  0.992749  ...  649.226514   
42104 -2.449294e-16        1.0 -0.120208  0.992749  ...  515.790171   
42105 -2.449294e-16        1.0 -0.120208  0.992749  ...

In [9]:
futr_exog_df = test.drop(["y"], axis=1)
futr_exog_df.head()
futr_exog_df.shape

(168, 33)

In [12]:

# ============================================================
# 1. Different AutoARIMA configurations to test
# ============================================================

param_grid = [
    # Small search space - faster
    {
        "season_length": 24,
        "max_p": 2,
        "max_q": 2,
        "max_P": 1,
        "max_Q": 1,
        "max_order": 4,
        "approximation": True
    },

    # Medium search space
    {
        "season_length": 24,
        "max_p": 3,
        "max_q": 3,
        "max_P": 1,
        "max_Q": 1,
        "max_order": 5,
        "approximation": True
    },

    {
        "season_length": 24,
        "max_p": 4,
        "max_q": 4,
        "max_P": 2,
        "max_Q": 2,
        "max_order": 6,
        "approximation": True
    },

    # Larger search space
    {
        "season_length": 24,
        "max_p": 5,
        "max_q": 5,
        "max_P": 2,
        "max_Q": 2,
        "max_order": 8,
        "approximation": False
    }
]


# ============================================================
# 2. Time-series cross-validation
# ============================================================

results = []

for i, params in enumerate(param_grid):

    print(f"\nRunning configuration {i+1}/{len(param_grid)}")
    print(params)

    # Create AutoARIMA using current parameters
    model = AutoARIMA(
        season_length=params["season_length"],
        max_p=params["max_p"],
        max_q=params["max_q"],
        max_P=params["max_P"],
        max_Q=params["max_Q"],
        max_order=params["max_order"],
        approximation=params["approximation"],
        alias="SARIMA"
    )

    # Create StatsForecast object
    sf = StatsForecast(
        models=[model],
        freq="H"
    )

    # --------------------------------------------------------
    # Cross-validation
    #
    # h=24      -> predict next 24 hours
    # n_windows=5 -> use 5 historical validation windows
    # step_size=24 -> move forward 24 hours each time
    # --------------------------------------------------------

    cv_df = sf.cross_validation(
        df=train,
        h=24,
        n_windows=5,
        step_size=24,
        refit=True
    )

    # --------------------------------------------------------
    # Calculate RMSE
    # --------------------------------------------------------

    rmse = np.sqrt(
        mean_squared_error(
            cv_df["y"],
            cv_df["SARIMA"]
        )
    )

    results.append({
        **params,
        "RMSE": rmse
    })

    print(f"RMSE = {rmse:.4f}")


# ============================================================
# 3. Show the best configuration
# ============================================================

results_df = pd.DataFrame(results)

results_df = results_df.sort_values(
    "RMSE",
    ascending=True
).reset_index(drop=True)

print("\n========== BEST CONFIGURATION ==========")
print(results_df.iloc[0])

print("\n========== ALL RESULTS ==========")
print(results_df)


Running configuration 1/4
{'season_length': 24, 'max_p': 2, 'max_q': 2, 'max_P': 1, 'max_Q': 1, 'max_order': 4, 'approximation': True}


d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\statsforecast\arima.py:672: UserWarning: possible convergence problem: minimize gave code 2]
  warnings.warn(
d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\statsforecast\arima.py:672: UserWarning: possible convergence problem: minimize gave code 1]
  warnings.warn(
d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\statsforecast\arima.py:672: UserWarning: possible convergence problem: minimize gave code 2]
  warnings.warn(
d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\statsforecast\arima.py:672: UserWarning: possible convergence problem: minimize gave code 2]
  warnings.warn(
d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\statsforecast\arima.py:672: UserWarning: possible convergence problem: minimize gave code 2]
  warnings.warn(
d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\statsforecast\arima.py:672: UserWarning: p

RMSE = 393.1144

Running configuration 2/4
{'season_length': 24, 'max_p': 3, 'max_q': 3, 'max_P': 1, 'max_Q': 1, 'max_order': 5, 'approximation': True}


d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\statsforecast\arima.py:672: UserWarning: possible convergence problem: minimize gave code 2]
  warnings.warn(
d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\statsforecast\arima.py:672: UserWarning: possible convergence problem: minimize gave code 1]
  warnings.warn(
d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\statsforecast\arima.py:672: UserWarning: possible convergence problem: minimize gave code 2]
  warnings.warn(
d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\statsforecast\arima.py:672: UserWarning: possible convergence problem: minimize gave code 2]
  warnings.warn(
d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\statsforecast\arima.py:672: UserWarning: possible convergence problem: minimize gave code 2]
  warnings.warn(
d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\statsforecast\arima.py:672: UserWarning: p

RMSE = 398.9561

Running configuration 3/4
{'season_length': 24, 'max_p': 4, 'max_q': 4, 'max_P': 2, 'max_Q': 2, 'max_order': 6, 'approximation': True}


d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\statsforecast\arima.py:672: UserWarning: possible convergence problem: minimize gave code 2]
  warnings.warn(
d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\statsforecast\arima.py:672: UserWarning: possible convergence problem: minimize gave code 1]
  warnings.warn(
d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\statsforecast\arima.py:672: UserWarning: possible convergence problem: minimize gave code 2]
  warnings.warn(
d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\statsforecast\arima.py:672: UserWarning: possible convergence problem: minimize gave code 2]
  warnings.warn(
d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\statsforecast\arima.py:672: UserWarning: possible convergence problem: minimize gave code 2]
  warnings.warn(
d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\statsforecast\arima.py:672: UserWarning: p

RMSE = 403.7564

Running configuration 4/4
{'season_length': 24, 'max_p': 5, 'max_q': 5, 'max_P': 2, 'max_Q': 2, 'max_order': 8, 'approximation': False}


KeyboardInterrupt: 

###
Running configuration 1/4 - RMSE = 393.1144

{'season_length': 24, 'max_p': 2, 'max_q': 2, 'max_P': 1, 'max_Q': 1, 'max_order': 4, 'approximation': True}


Running configuration 2/4 - RMSE = 398.9561

{'season_length': 24, 'max_p': 3, 'max_q': 3, 'max_P': 1, 'max_Q': 1, 'max_order': 5, 'approximation': True}


Running configuration 3/4 - RMSE = 403.7564

{'season_length': 24, 'max_p': 4, 'max_q': 4, 'max_P': 2, 'max_Q': 2, 'max_order': 6, 'approximation': True}

###

In [13]:
# Get best parameters
best = results_df.iloc[0]

print("Best parameters:")
print(best)


# ------------------------------------------------------------
# Train final model using the best configuration
# ------------------------------------------------------------

final_model = AutoARIMA(
    season_length=int(best["season_length"]),
    max_p=int(best["max_p"]),
    max_q=int(best["max_q"]),
    max_P=int(best["max_P"]),
    max_Q=int(best["max_Q"]),
    max_order=int(best["max_order"]),
    approximation=bool(best["approximation"]),
    alias="SARIMA_exog"
)

sf_final = StatsForecast(
    models=[final_model],
    freq="H"
)

sf_final.fit(df=train)

NameError: name 'results_df' is not defined

In [14]:
from sklearn.ensemble import GradientBoostingRegressor

In [15]:
print("Loading data...")
df = pd.read_csv('D:\\Users\\Md Mahfuzur Rahman\\Desktop\\Research\\for_me\\cleaned_combined_data.csv')

df['datetime'] = pd.to_datetime(df['datetime'])
df = df.sort_values('datetime').reset_index(drop=True)

Loading data...


In [16]:
df['hour'] = df['datetime'].dt.hour
df['dayofweek'] = df['datetime'].dt.dayofweek
df['month'] = df['datetime'].dt.month
df['dayofyear'] = df['datetime'].dt.dayofyear

# -----------------------------
# Cyclic time features
# -----------------------------

df['hour_sin'] = np.sin(2 * np.pi * df['hour'] / 24)
df['hour_cos'] = np.cos(2 * np.pi * df['hour'] / 24)

df['dow_sin'] = np.sin(2 * np.pi * df['dayofweek'] / 7)
df['dow_cos'] = np.cos(2 * np.pi * df['dayofweek'] / 7)

df['month_sin'] = np.sin(2 * np.pi * df['month'] / 12)
df['month_cos'] = np.cos(2 * np.pi * df['month'] / 12)

df['doy_sin'] = np.sin(2 * np.pi * df['dayofyear'] / 365)
df['doy_cos'] = np.cos(2 * np.pi * df['dayofyear'] / 365)


# -----------------------------
# Lag features
# -----------------------------

for lag in [1, 2, 3, 6, 12, 24, 48, 72, 168, 336]:
    df[f'lag{lag}'] = df['demand_mw'].shift(lag)


# -----------------------------
# Rolling features
# IMPORTANT: shift(1)
# prevents target leakage
# -----------------------------

past_demand = df['demand_mw'].shift(1)

for window in [6, 12, 24, 48, 168]:

    df[f'roll{window}_mean'] = (
        past_demand.rolling(window).mean()
    )

    df[f'roll{window}_std'] = (
        past_demand.rolling(window).std()
    )


# Remove NaN
df = df.dropna().reset_index(drop=True)

In [17]:
df.shape

(42275, 37)

In [18]:
FEATURES = [
    'hour_sin','hour_cos','dow_sin','dow_cos','month_sin','month_cos','doy_sin','doy_cos',

    'load_shedding','temp_mean','wspd_mean',

    'lag1','lag2','lag3','lag6','lag12','lag24','lag48','lag72','lag168','lag336',

    'roll6_mean','roll6_std','roll12_mean','roll12_std','roll24_mean','roll24_std',
    'roll48_mean','roll48_std','roll168_mean','roll168_std'
]

TARGET = 'demand_mw'

In [19]:
from sklearn.model_selection import train_test_split
X = df[FEATURES]
y = df[TARGET]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, shuffle=False)

# Reset indices for easier access
X_train = X_train.reset_index(drop=True)
X_test = X_test.reset_index(drop=True)
y_train = y_train.reset_index(drop=True)
y_test = y_test.reset_index(drop=True)

# Create aliases for backward compatibility
X_tr, X_te = X_train, X_test
y_tr, y_te = y_train, y_test

print(f"Train rows: {len(X_train):,}   Test rows: {len(X_test):,}")

Train rows: 33,820   Test rows: 8,455


In [20]:
from sklearn.preprocessing import StandardScaler
import torch

X_scaler = StandardScaler()

X_tr_sc = X_scaler.fit_transform(X_tr)
X_te_sc = X_scaler.transform(X_te)

# PyTorch tensors
X_tr = torch.tensor(X_tr_sc, dtype=torch.float32)
X_te = torch.tensor(X_te_sc, dtype=torch.float32)

y_tr = torch.tensor(y_tr, dtype=torch.float32).reshape(-1, 1)
y_te = torch.tensor(y_te, dtype=torch.float32).reshape(-1, 1)

In [23]:
import optuna
import xgboost as xgb
import numpy as np

from sklearn.metrics import mean_squared_error


# ============================================================
# 1. Create a time-based validation split
# ============================================================
# IMPORTANT:
# We use only X_tr / y_tr for Optuna.
# X_te / y_te stays completely untouched.

val_size = int(len(X_tr) * 0.20)

# Works with both pandas DataFrame/Series and NumPy arrays
if hasattr(X_tr, "iloc"):
    X_train_cv = X_tr.iloc[:-val_size]
    X_val_cv   = X_tr.iloc[-val_size:]

    y_train_cv = y_tr.iloc[:-val_size]
    y_val_cv   = y_tr.iloc[-val_size:]
else:
    X_train_cv = X_tr[:-val_size]
    X_val_cv   = X_tr[-val_size:]

    y_train_cv = y_tr[:-val_size]
    y_val_cv   = y_tr[-val_size:]


# ============================================================
# 2. Optuna objective function
# ============================================================

def objective(trial):

    # --------------------------------------------------------
    # Optuna chooses these hyperparameters automatically
    # --------------------------------------------------------

    params = {
        "n_estimators": trial.suggest_int(
            "n_estimators", 300, 1200, step=100
        ),

        "learning_rate": trial.suggest_float(
            "learning_rate", 0.01, 0.10, log=True
        ),

        "max_depth": trial.suggest_int(
            "max_depth", 3, 8
        ),

        "min_child_weight": trial.suggest_int(
            "min_child_weight", 1, 15
        ),

        "subsample": trial.suggest_float(
            "subsample", 0.70, 1.0
        ),

        "colsample_bytree": trial.suggest_float(
            "colsample_bytree", 0.70, 1.0
        ),

        "gamma": trial.suggest_float(
            "gamma", 0.0, 1.0
        ),

        "reg_alpha": trial.suggest_float(
            "reg_alpha", 1e-8, 1.0, log=True
        ),

        "reg_lambda": trial.suggest_float(
            "reg_lambda", 1.0, 10.0, log=True
        )
    }


    # --------------------------------------------------------
    # Create XGBoost model
    # --------------------------------------------------------

    model = xgb.XGBRegressor(
        objective="reg:squarederror",

        **params,

        random_state=42,
        tree_method="hist",
        n_jobs=-1
    )


    # --------------------------------------------------------
    # Train ONLY on earlier data
    # --------------------------------------------------------

    model.fit(
        X_train_cv,
        y_train_cv,
        verbose=False
    )


    # --------------------------------------------------------
    # Predict validation period
    # --------------------------------------------------------

    pred = model.predict(X_val_cv)


    # --------------------------------------------------------
    # Calculate RMSE
    # --------------------------------------------------------

    rmse = np.sqrt(
        mean_squared_error(y_val_cv, pred)
    )

    return rmse


# ============================================================
# 3. Create Optuna study
# ============================================================

study = optuna.create_study(
    direction="minimize",
    study_name="XGBoost_Electricity_Demand"
)


# ============================================================
# 4. Start hyperparameter optimization
# ============================================================

study.optimize(
    objective,
    n_trials=50,       # Try 50 different combinations
    show_progress_bar=True
)


# ============================================================
# 5. Display best parameters
# ============================================================

print("\n==========================================")
print("BEST XGBOOST PARAMETERS")
print("==========================================")

print("Best RMSE:")
print(study.best_value)

print("\nBest Parameters:")
for key, value in study.best_params.items():
    print(f"{key}: {value}")

[I 2026-08-22 01:08:50,295] A new study created in memory with name: XGBoost_Electricity_Demand
Best trial: 0. Best value: 336.043:   2%|▏         | 1/50 [00:01<00:57,  1.17s/it]

[I 2026-08-22 01:08:51,467] Trial 0 finished with value: 336.04301249988816 and parameters: {'n_estimators': 300, 'learning_rate': 0.09100358391708445, 'max_depth': 4, 'min_child_weight': 13, 'subsample': 0.8266917415465012, 'colsample_bytree': 0.835413985646015, 'gamma': 0.6126005677462565, 'reg_alpha': 1.4890087037719996e-05, 'reg_lambda': 2.4026501357231327}. Best is trial 0 with value: 336.04301249988816.


Best trial: 1. Best value: 335.989:   4%|▍         | 2/50 [00:04<02:08,  2.69s/it]

[I 2026-08-22 01:08:55,217] Trial 1 finished with value: 335.9890134922569 and parameters: {'n_estimators': 1200, 'learning_rate': 0.010771421158631457, 'max_depth': 5, 'min_child_weight': 8, 'subsample': 0.8481172195204313, 'colsample_bytree': 0.8077821039021381, 'gamma': 0.5067734583747873, 'reg_alpha': 1.8623306316887333e-06, 'reg_lambda': 4.166727524482641}. Best is trial 1 with value: 335.9890134922569.


Best trial: 1. Best value: 335.989:   6%|▌         | 3/50 [00:05<01:25,  1.82s/it]

[I 2026-08-22 01:08:56,010] Trial 2 finished with value: 403.020451559223 and parameters: {'n_estimators': 400, 'learning_rate': 0.013167509477385647, 'max_depth': 4, 'min_child_weight': 13, 'subsample': 0.8480009553040236, 'colsample_bytree': 0.878766569863299, 'gamma': 0.9712952539442075, 'reg_alpha': 1.4074878085191043e-08, 'reg_lambda': 1.071413359514702}. Best is trial 1 with value: 335.9890134922569.


Best trial: 1. Best value: 335.989:   8%|▊         | 4/50 [00:06<01:08,  1.49s/it]

[I 2026-08-22 01:08:56,998] Trial 3 finished with value: 337.0130862629818 and parameters: {'n_estimators': 300, 'learning_rate': 0.07150925918157704, 'max_depth': 6, 'min_child_weight': 7, 'subsample': 0.8156195082920333, 'colsample_bytree': 0.9097643581858085, 'gamma': 0.21077275249472405, 'reg_alpha': 1.81701358625827e-05, 'reg_lambda': 3.64560376881404}. Best is trial 1 with value: 335.9890134922569.


Best trial: 1. Best value: 335.989:  10%|█         | 5/50 [00:09<01:30,  2.01s/it]

[I 2026-08-22 01:08:59,938] Trial 4 finished with value: 345.0673168020988 and parameters: {'n_estimators': 800, 'learning_rate': 0.09341861725332597, 'max_depth': 6, 'min_child_weight': 2, 'subsample': 0.9770220132855365, 'colsample_bytree': 0.9299505087829174, 'gamma': 0.8792224955035517, 'reg_alpha': 4.517257004741083e-05, 'reg_lambda': 7.897536472344565}. Best is trial 1 with value: 335.9890134922569.


Best trial: 1. Best value: 335.989:  12%|█▏        | 6/50 [00:12<01:45,  2.39s/it]

[I 2026-08-22 01:09:03,049] Trial 5 finished with value: 339.26796194158976 and parameters: {'n_estimators': 1100, 'learning_rate': 0.049825315703153635, 'max_depth': 5, 'min_child_weight': 7, 'subsample': 0.7919126967650862, 'colsample_bytree': 0.8823911202404544, 'gamma': 0.7052258813010278, 'reg_alpha': 0.04541316674912676, 'reg_lambda': 6.411402343391562}. Best is trial 1 with value: 335.9890134922569.


Best trial: 1. Best value: 335.989:  14%|█▍        | 7/50 [00:16<02:02,  2.84s/it]

[I 2026-08-22 01:09:06,822] Trial 6 finished with value: 351.11948197871334 and parameters: {'n_estimators': 1100, 'learning_rate': 0.05073295907796375, 'max_depth': 6, 'min_child_weight': 3, 'subsample': 0.7157649797403757, 'colsample_bytree': 0.8658430480969727, 'gamma': 0.8143750402556358, 'reg_alpha': 3.414524529731325e-05, 'reg_lambda': 2.1233263247852494}. Best is trial 1 with value: 335.9890134922569.


Best trial: 1. Best value: 335.989:  16%|█▌        | 8/50 [00:17<01:37,  2.32s/it]

[I 2026-08-22 01:09:08,021] Trial 7 finished with value: 341.11617110230935 and parameters: {'n_estimators': 600, 'learning_rate': 0.02470761355703964, 'max_depth': 4, 'min_child_weight': 10, 'subsample': 0.9090399364355667, 'colsample_bytree': 0.8628521665883044, 'gamma': 0.4229137036932975, 'reg_alpha': 5.9948846119213325e-05, 'reg_lambda': 2.3919852957535093}. Best is trial 1 with value: 335.9890134922569.


Best trial: 8. Best value: 335.578:  18%|█▊        | 9/50 [00:20<01:36,  2.36s/it]

[I 2026-08-22 01:09:10,460] Trial 8 finished with value: 335.5777084931298 and parameters: {'n_estimators': 1200, 'learning_rate': 0.048553089749432385, 'max_depth': 4, 'min_child_weight': 15, 'subsample': 0.7276844072519013, 'colsample_bytree': 0.7459764355371753, 'gamma': 0.7745230520611334, 'reg_alpha': 0.007550087156066418, 'reg_lambda': 1.559201639998125}. Best is trial 8 with value: 335.5777084931298.


Best trial: 8. Best value: 335.578:  20%|██        | 10/50 [00:25<02:13,  3.35s/it]

[I 2026-08-22 01:09:15,995] Trial 9 finished with value: 353.2465388188538 and parameters: {'n_estimators': 800, 'learning_rate': 0.08189951733252475, 'max_depth': 7, 'min_child_weight': 5, 'subsample': 0.7142442283230789, 'colsample_bytree': 0.981032254554131, 'gamma': 0.018958411092207084, 'reg_alpha': 4.8950935407735035e-06, 'reg_lambda': 2.8126493470967717}. Best is trial 8 with value: 335.5777084931298.


Best trial: 8. Best value: 335.578:  22%|██▏       | 11/50 [00:33<03:01,  4.66s/it]

[I 2026-08-22 01:09:23,659] Trial 10 finished with value: 346.7682718689817 and parameters: {'n_estimators': 900, 'learning_rate': 0.02309491213453989, 'max_depth': 8, 'min_child_weight': 15, 'subsample': 0.9950925953122671, 'colsample_bytree': 0.7006068019363655, 'gamma': 0.28737088739884004, 'reg_alpha': 0.35356547148451334, 'reg_lambda': 1.170030654998897}. Best is trial 8 with value: 335.5777084931298.


Best trial: 8. Best value: 335.578:  24%|██▍       | 12/50 [00:35<02:29,  3.94s/it]

[I 2026-08-22 01:09:25,947] Trial 11 finished with value: 391.0742616498815 and parameters: {'n_estimators': 1200, 'learning_rate': 0.010174406840297582, 'max_depth': 3, 'min_child_weight': 9, 'subsample': 0.8922421727964929, 'colsample_bytree': 0.7664844841511962, 'gamma': 0.539852570833163, 'reg_alpha': 0.0018002456350594935, 'reg_lambda': 5.003645117407974}. Best is trial 8 with value: 335.5777084931298.


Best trial: 8. Best value: 335.578:  26%|██▌       | 13/50 [00:38<02:08,  3.48s/it]

[I 2026-08-22 01:09:28,381] Trial 12 finished with value: 350.53911604127717 and parameters: {'n_estimators': 1200, 'learning_rate': 0.02264124005220033, 'max_depth': 3, 'min_child_weight': 11, 'subsample': 0.762509636175758, 'colsample_bytree': 0.7825950728616407, 'gamma': 0.4581087658995419, 'reg_alpha': 7.70328657411447e-08, 'reg_lambda': 1.610600994371195}. Best is trial 8 with value: 335.5777084931298.


Best trial: 8. Best value: 335.578:  28%|██▊       | 14/50 [00:41<02:01,  3.36s/it]

[I 2026-08-22 01:09:31,471] Trial 13 finished with value: 338.6362239741933 and parameters: {'n_estimators': 1000, 'learning_rate': 0.016029145888708678, 'max_depth': 5, 'min_child_weight': 14, 'subsample': 0.7560218167881307, 'colsample_bytree': 0.7645075740726056, 'gamma': 0.6910386315075617, 'reg_alpha': 4.760917043256403e-07, 'reg_lambda': 9.661104041802284}. Best is trial 8 with value: 335.5777084931298.


Best trial: 8. Best value: 335.578:  30%|███       | 15/50 [00:44<02:02,  3.50s/it]

[I 2026-08-22 01:09:35,287] Trial 14 finished with value: 345.29578579458513 and parameters: {'n_estimators': 1200, 'learning_rate': 0.04457570049542342, 'max_depth': 5, 'min_child_weight': 7, 'subsample': 0.8849943266030695, 'colsample_bytree': 0.7165045688850186, 'gamma': 0.7956869944134926, 'reg_alpha': 0.002387458032436245, 'reg_lambda': 4.200748459594954}. Best is trial 8 with value: 335.5777084931298.


Best trial: 8. Best value: 335.578:  32%|███▏      | 16/50 [00:47<01:45,  3.10s/it]

[I 2026-08-22 01:09:37,458] Trial 15 finished with value: 336.59290136602704 and parameters: {'n_estimators': 1000, 'learning_rate': 0.03712256644174027, 'max_depth': 4, 'min_child_weight': 11, 'subsample': 0.9460709751263491, 'colsample_bytree': 0.8030357719641137, 'gamma': 0.3616875929478038, 'reg_alpha': 0.003500998807046559, 'reg_lambda': 1.5746954955103676}. Best is trial 8 with value: 335.5777084931298.


Best trial: 8. Best value: 335.578:  34%|███▍      | 17/50 [00:50<01:41,  3.09s/it]

[I 2026-08-22 01:09:40,519] Trial 16 finished with value: 343.24186306815784 and parameters: {'n_estimators': 1000, 'learning_rate': 0.03170390949865751, 'max_depth': 5, 'min_child_weight': 1, 'subsample': 0.7568719947735135, 'colsample_bytree': 0.740944956163409, 'gamma': 0.996813187488198, 'reg_alpha': 0.631009260296799, 'reg_lambda': 3.2824140262052186}. Best is trial 8 with value: 335.5777084931298.


Best trial: 8. Best value: 335.578:  36%|███▌      | 18/50 [00:52<01:28,  2.76s/it]

[I 2026-08-22 01:09:42,503] Trial 17 finished with value: 365.20197750696803 and parameters: {'n_estimators': 1100, 'learning_rate': 0.01687592312261526, 'max_depth': 3, 'min_child_weight': 4, 'subsample': 0.8606709578048628, 'colsample_bytree': 0.8181794001116096, 'gamma': 0.5874543755971529, 'reg_alpha': 9.84063588354507e-07, 'reg_lambda': 5.72616739934938}. Best is trial 8 with value: 335.5777084931298.


Best trial: 8. Best value: 335.578:  38%|███▊      | 19/50 [00:56<01:38,  3.18s/it]

[I 2026-08-22 01:09:46,688] Trial 18 finished with value: 354.9302418257706 and parameters: {'n_estimators': 800, 'learning_rate': 0.05871809472766556, 'max_depth': 7, 'min_child_weight': 9, 'subsample': 0.9385350597627662, 'colsample_bytree': 0.734208922830693, 'gamma': 0.08725865164756713, 'reg_alpha': 0.000292901860666858, 'reg_lambda': 1.6583433157858367}. Best is trial 8 with value: 335.5777084931298.


Best trial: 8. Best value: 335.578:  40%|████      | 20/50 [00:57<01:18,  2.61s/it]

[I 2026-08-22 01:09:47,947] Trial 19 finished with value: 396.778708791185 and parameters: {'n_estimators': 600, 'learning_rate': 0.010229579838084075, 'max_depth': 4, 'min_child_weight': 5, 'subsample': 0.7930907481234714, 'colsample_bytree': 0.7975123657672497, 'gamma': 0.6896978938549434, 'reg_alpha': 0.03612737310095001, 'reg_lambda': 4.519299721709066}. Best is trial 8 with value: 335.5777084931298.


Best trial: 20. Best value: 333.754:  42%|████▏     | 21/50 [01:01<01:25,  2.96s/it]

[I 2026-08-22 01:09:51,715] Trial 20 finished with value: 333.75428368037467 and parameters: {'n_estimators': 1200, 'learning_rate': 0.029819249042360472, 'max_depth': 5, 'min_child_weight': 12, 'subsample': 0.7093363500038761, 'colsample_bytree': 0.8329194196237018, 'gamma': 0.46725564702775446, 'reg_alpha': 5.712813104847111e-07, 'reg_lambda': 1.2908855833355077}. Best is trial 20 with value: 333.75428368037467.


Best trial: 21. Best value: 333.327:  44%|████▍     | 22/50 [01:05<01:31,  3.26s/it]

[I 2026-08-22 01:09:55,683] Trial 21 finished with value: 333.32716660962393 and parameters: {'n_estimators': 1200, 'learning_rate': 0.032292376898858006, 'max_depth': 5, 'min_child_weight': 15, 'subsample': 0.7050286914541304, 'colsample_bytree': 0.8351955557300216, 'gamma': 0.4995783936835618, 'reg_alpha': 8.190085137983557e-07, 'reg_lambda': 1.3003985662928184}. Best is trial 21 with value: 333.32716660962393.


Best trial: 21. Best value: 333.327:  46%|████▌     | 23/50 [01:09<01:31,  3.40s/it]

[I 2026-08-22 01:09:59,414] Trial 22 finished with value: 334.1001323518145 and parameters: {'n_estimators': 1100, 'learning_rate': 0.030440070674317116, 'max_depth': 6, 'min_child_weight': 15, 'subsample': 0.702072461070605, 'colsample_bytree': 0.8436187152270064, 'gamma': 0.34535600045482745, 'reg_alpha': 1.0214384516863981e-07, 'reg_lambda': 1.2930876911633917}. Best is trial 21 with value: 333.32716660962393.


Best trial: 23. Best value: 332.451:  48%|████▊     | 24/50 [01:12<01:30,  3.48s/it]

[I 2026-08-22 01:10:03,033] Trial 23 finished with value: 332.45084223084774 and parameters: {'n_estimators': 1100, 'learning_rate': 0.03333290608993889, 'max_depth': 6, 'min_child_weight': 13, 'subsample': 0.7028308546841446, 'colsample_bytree': 0.8401504222082523, 'gamma': 0.31066982124277376, 'reg_alpha': 1.2404424953989102e-07, 'reg_lambda': 1.290571487746729}. Best is trial 23 with value: 332.45084223084774.


Best trial: 23. Best value: 332.451:  50%|█████     | 25/50 [01:17<01:35,  3.84s/it]

[I 2026-08-22 01:10:07,756] Trial 24 finished with value: 338.4054289975266 and parameters: {'n_estimators': 900, 'learning_rate': 0.036483821414967595, 'max_depth': 7, 'min_child_weight': 12, 'subsample': 0.7332557294695984, 'colsample_bytree': 0.9075699079418273, 'gamma': 0.21360316353342423, 'reg_alpha': 1.5894387882108476e-07, 'reg_lambda': 1.0108539163155807}. Best is trial 23 with value: 332.45084223084774.


Best trial: 23. Best value: 332.451:  52%|█████▏    | 26/50 [01:21<01:30,  3.76s/it]

[I 2026-08-22 01:10:11,346] Trial 25 finished with value: 338.5748970870404 and parameters: {'n_estimators': 1000, 'learning_rate': 0.028390988938395166, 'max_depth': 6, 'min_child_weight': 13, 'subsample': 0.7422567442653856, 'colsample_bytree': 0.8344297066206791, 'gamma': 0.23173037422769316, 'reg_alpha': 1.4214511979106363e-08, 'reg_lambda': 1.3873660547401336}. Best is trial 23 with value: 332.45084223084774.


Best trial: 26. Best value: 329.862:  54%|█████▍    | 27/50 [01:24<01:27,  3.80s/it]

[I 2026-08-22 01:10:15,218] Trial 26 finished with value: 329.8622226589459 and parameters: {'n_estimators': 1100, 'learning_rate': 0.018921391656950854, 'max_depth': 5, 'min_child_weight': 12, 'subsample': 0.7060090278704589, 'colsample_bytree': 0.9452367426083146, 'gamma': 0.37122128460141446, 'reg_alpha': 2.8763150908296133e-07, 'reg_lambda': 1.792197076270558}. Best is trial 26 with value: 329.8622226589459.


Best trial: 26. Best value: 329.862:  56%|█████▌    | 28/50 [01:30<01:38,  4.47s/it]

[I 2026-08-22 01:10:21,273] Trial 27 finished with value: 333.983368785034 and parameters: {'n_estimators': 900, 'learning_rate': 0.019497924620722796, 'max_depth': 8, 'min_child_weight': 14, 'subsample': 0.7836627133873862, 'colsample_bytree': 0.9957902899221382, 'gamma': 0.1368547797261494, 'reg_alpha': 3.2525576229904714e-06, 'reg_lambda': 1.934458666148004}. Best is trial 26 with value: 329.8622226589459.


Best trial: 26. Best value: 329.862:  58%|█████▊    | 29/50 [01:36<01:40,  4.81s/it]

[I 2026-08-22 01:10:26,863] Trial 28 finished with value: 340.10633677204544 and parameters: {'n_estimators': 1100, 'learning_rate': 0.03893708571485481, 'max_depth': 7, 'min_child_weight': 11, 'subsample': 0.700704165248967, 'colsample_bytree': 0.9436643208111465, 'gamma': 0.4070895133186875, 'reg_alpha': 4.823841950966811e-08, 'reg_lambda': 1.9612157062361109}. Best is trial 26 with value: 329.8622226589459.


Best trial: 29. Best value: 326.904:  60%|██████    | 30/50 [01:41<01:36,  4.81s/it]

[I 2026-08-22 01:10:31,686] Trial 29 finished with value: 326.9043485868611 and parameters: {'n_estimators': 1000, 'learning_rate': 0.019926632702212937, 'max_depth': 6, 'min_child_weight': 13, 'subsample': 0.7301739609011664, 'colsample_bytree': 0.9678283669065567, 'gamma': 0.31503543253102495, 'reg_alpha': 2.8401633828131725e-07, 'reg_lambda': 2.4203926789071084}. Best is trial 29 with value: 326.9043485868611.


Best trial: 29. Best value: 326.904:  62%|██████▏   | 31/50 [01:44<01:20,  4.21s/it]

[I 2026-08-22 01:10:34,500] Trial 30 finished with value: 330.3596406758852 and parameters: {'n_estimators': 700, 'learning_rate': 0.01418370065365392, 'max_depth': 6, 'min_child_weight': 13, 'subsample': 0.7671560173704507, 'colsample_bytree': 0.964086508471889, 'gamma': 0.3088230264517027, 'reg_alpha': 2.6918057051355765e-07, 'reg_lambda': 2.774300332617024}. Best is trial 29 with value: 326.9043485868611.


Best trial: 29. Best value: 326.904:  64%|██████▍   | 32/50 [01:46<01:05,  3.64s/it]

[I 2026-08-22 01:10:36,797] Trial 31 finished with value: 333.3368580542812 and parameters: {'n_estimators': 600, 'learning_rate': 0.013857875841246137, 'max_depth': 6, 'min_child_weight': 13, 'subsample': 0.772256120730928, 'colsample_bytree': 0.9629886557549078, 'gamma': 0.29924704189343243, 'reg_alpha': 2.3490654151682618e-07, 'reg_lambda': 2.6509831378188866}. Best is trial 29 with value: 326.9043485868611.


Best trial: 29. Best value: 326.904:  66%|██████▌   | 33/50 [01:49<00:58,  3.45s/it]

[I 2026-08-22 01:10:39,803] Trial 32 finished with value: 329.39065427695425 and parameters: {'n_estimators': 700, 'learning_rate': 0.01882479670270877, 'max_depth': 6, 'min_child_weight': 12, 'subsample': 0.7384600929104715, 'colsample_bytree': 0.957849883054375, 'gamma': 0.277032455116868, 'reg_alpha': 2.9014332852463852e-08, 'reg_lambda': 3.0594612359950304}. Best is trial 29 with value: 326.9043485868611.


Best trial: 29. Best value: 326.904:  68%|██████▊   | 34/50 [01:53<00:59,  3.70s/it]

[I 2026-08-22 01:10:44,087] Trial 33 finished with value: 331.7173472061719 and parameters: {'n_estimators': 700, 'learning_rate': 0.019072678662606046, 'max_depth': 7, 'min_child_weight': 12, 'subsample': 0.7463197744365605, 'colsample_bytree': 0.9588985245084939, 'gamma': 0.1124394707657701, 'reg_alpha': 3.0986928384075346e-08, 'reg_lambda': 3.031464445981658}. Best is trial 29 with value: 326.9043485868611.


Best trial: 29. Best value: 326.904:  70%|███████   | 35/50 [01:56<00:48,  3.26s/it]

[I 2026-08-22 01:10:46,337] Trial 34 finished with value: 344.88842172143154 and parameters: {'n_estimators': 500, 'learning_rate': 0.013224006749824227, 'max_depth': 6, 'min_child_weight': 10, 'subsample': 0.8086991749362311, 'colsample_bytree': 0.9300895819464413, 'gamma': 0.1692575524926367, 'reg_alpha': 5.710979272393443e-06, 'reg_lambda': 2.378445123548832}. Best is trial 29 with value: 326.9043485868611.


Best trial: 29. Best value: 326.904:  72%|███████▏  | 36/50 [01:57<00:39,  2.85s/it]

[I 2026-08-22 01:10:48,173] Trial 35 finished with value: 364.02933861846907 and parameters: {'n_estimators': 400, 'learning_rate': 0.012165087561463579, 'max_depth': 6, 'min_child_weight': 14, 'subsample': 0.734124611041968, 'colsample_bytree': 0.966560377426292, 'gamma': 0.2541794423979589, 'reg_alpha': 2.129565123354582e-08, 'reg_lambda': 3.4639437953149477}. Best is trial 29 with value: 326.9043485868611.


Best trial: 29. Best value: 326.904:  74%|███████▍  | 37/50 [02:00<00:34,  2.65s/it]

[I 2026-08-22 01:10:50,414] Trial 36 finished with value: 331.4863072209771 and parameters: {'n_estimators': 700, 'learning_rate': 0.016470538684967697, 'max_depth': 5, 'min_child_weight': 12, 'subsample': 0.7695143136237147, 'colsample_bytree': 0.9972792305560567, 'gamma': 0.3833985724693798, 'reg_alpha': 1.9050335917316164e-06, 'reg_lambda': 2.0878277177901063}. Best is trial 29 with value: 326.9043485868611.


Best trial: 29. Best value: 326.904:  76%|███████▌  | 38/50 [02:02<00:32,  2.68s/it]

[I 2026-08-22 01:10:53,152] Trial 37 finished with value: 331.52073841616607 and parameters: {'n_estimators': 700, 'learning_rate': 0.021255256560881022, 'max_depth': 6, 'min_child_weight': 10, 'subsample': 0.8371152721638714, 'colsample_bytree': 0.9217595688574244, 'gamma': 0.18298039071262046, 'reg_alpha': 3.139992921258495e-07, 'reg_lambda': 3.736213826492824}. Best is trial 29 with value: 326.9043485868611.


Best trial: 29. Best value: 326.904:  78%|███████▊  | 39/50 [02:05<00:29,  2.72s/it]

[I 2026-08-22 01:10:55,972] Trial 38 finished with value: 339.09591425214785 and parameters: {'n_estimators': 500, 'learning_rate': 0.015037111253465475, 'max_depth': 7, 'min_child_weight': 14, 'subsample': 0.805979523304175, 'colsample_bytree': 0.901025716551794, 'gamma': 0.3466847369251933, 'reg_alpha': 1.4515453086659756e-05, 'reg_lambda': 2.7002480622806337}. Best is trial 29 with value: 326.9043485868611.


Best trial: 39. Best value: 326.894:  80%|████████  | 40/50 [02:07<00:25,  2.51s/it]

[I 2026-08-22 01:10:57,978] Trial 39 finished with value: 326.8935821678058 and parameters: {'n_estimators': 800, 'learning_rate': 0.026340059352236846, 'max_depth': 5, 'min_child_weight': 11, 'subsample': 0.7263709147024497, 'colsample_bytree': 0.9493553396095371, 'gamma': 0.03818554358601167, 'reg_alpha': 4.0561758823329404e-08, 'reg_lambda': 2.310596649947486}. Best is trial 39 with value: 326.8935821678058.


Best trial: 39. Best value: 326.894:  82%|████████▏ | 41/50 [02:09<00:21,  2.43s/it]

[I 2026-08-22 01:11:00,214] Trial 40 finished with value: 329.8104450855067 and parameters: {'n_estimators': 900, 'learning_rate': 0.025752021197883255, 'max_depth': 5, 'min_child_weight': 8, 'subsample': 0.720703460307641, 'colsample_bytree': 0.9419274945062309, 'gamma': 0.04228937372490549, 'reg_alpha': 1.2814392620660668e-08, 'reg_lambda': 2.3008894878550836}. Best is trial 39 with value: 326.8935821678058.


Best trial: 39. Best value: 326.894:  84%|████████▍ | 42/50 [02:12<00:19,  2.43s/it]

[I 2026-08-22 01:11:02,648] Trial 41 finished with value: 331.3722062311503 and parameters: {'n_estimators': 900, 'learning_rate': 0.02522579162142814, 'max_depth': 5, 'min_child_weight': 8, 'subsample': 0.719741793821277, 'colsample_bytree': 0.9454013724314365, 'gamma': 0.005449611233108227, 'reg_alpha': 1.0327907166145298e-08, 'reg_lambda': 2.343158732332549}. Best is trial 39 with value: 326.8935821678058.


Best trial: 39. Best value: 326.894:  86%|████████▌ | 43/50 [02:14<00:16,  2.36s/it]

[I 2026-08-22 01:11:04,840] Trial 42 finished with value: 330.46288558626367 and parameters: {'n_estimators': 800, 'learning_rate': 0.019070589719083333, 'max_depth': 5, 'min_child_weight': 9, 'subsample': 0.7221469747954856, 'colsample_bytree': 0.9811954776465533, 'gamma': 0.0667182067033195, 'reg_alpha': 2.6998725493216966e-08, 'reg_lambda': 1.8508091517669847}. Best is trial 39 with value: 326.8935821678058.


Best trial: 39. Best value: 326.894:  88%|████████▊ | 44/50 [02:16<00:13,  2.26s/it]

[I 2026-08-22 01:11:06,879] Trial 43 finished with value: 334.86309002710345 and parameters: {'n_estimators': 900, 'learning_rate': 0.026242638949295332, 'max_depth': 4, 'min_child_weight': 6, 'subsample': 0.7445685565621316, 'colsample_bytree': 0.9405854418971566, 'gamma': 0.033666702634427365, 'reg_alpha': 4.848678688929397e-08, 'reg_lambda': 2.263380967195444}. Best is trial 39 with value: 326.8935821678058.


Best trial: 39. Best value: 326.894:  90%|█████████ | 45/50 [02:18<00:11,  2.23s/it]

[I 2026-08-22 01:11:09,046] Trial 44 finished with value: 329.8159050894908 and parameters: {'n_estimators': 800, 'learning_rate': 0.021392321088044803, 'max_depth': 5, 'min_child_weight': 11, 'subsample': 0.7285263637573844, 'colsample_bytree': 0.8913334881809606, 'gamma': 0.25411518312489545, 'reg_alpha': 6.581102580565023e-08, 'reg_lambda': 1.7978873292052755}. Best is trial 39 with value: 326.8935821678058.


Best trial: 39. Best value: 326.894:  92%|█████████▏| 46/50 [02:20<00:08,  2.11s/it]

[I 2026-08-22 01:11:10,871] Trial 45 finished with value: 334.84356188748797 and parameters: {'n_estimators': 800, 'learning_rate': 0.02228132683911036, 'max_depth': 4, 'min_child_weight': 8, 'subsample': 0.7282021302439238, 'colsample_bytree': 0.8913859178889146, 'gamma': 0.14081629792231443, 'reg_alpha': 5.4947489771984514e-08, 'reg_lambda': 2.590024863508746}. Best is trial 39 with value: 326.8935821678058.


Best trial: 39. Best value: 326.894:  94%|█████████▍| 47/50 [02:22<00:06,  2.14s/it]

[I 2026-08-22 01:11:13,063] Trial 46 finished with value: 328.7006022773612 and parameters: {'n_estimators': 800, 'learning_rate': 0.02592127796486521, 'max_depth': 5, 'min_child_weight': 11, 'subsample': 0.7443224078315883, 'colsample_bytree': 0.9230372286725055, 'gamma': 0.0556157687655392, 'reg_alpha': 1.5242171899219322e-08, 'reg_lambda': 3.1425579014151683}. Best is trial 39 with value: 326.8935821678058.


Best trial: 39. Best value: 326.894:  96%|█████████▌| 48/50 [02:28<00:06,  3.07s/it]

[I 2026-08-22 01:11:18,179] Trial 47 finished with value: 335.18054836162554 and parameters: {'n_estimators': 900, 'learning_rate': 0.02729400050970018, 'max_depth': 6, 'min_child_weight': 10, 'subsample': 0.7833858977861292, 'colsample_bytree': 0.9170427346352695, 'gamma': 0.05976623290073155, 'reg_alpha': 1.250378770094296e-08, 'reg_lambda': 3.14030410092545}. Best is trial 39 with value: 326.8935821678058.


Best trial: 39. Best value: 326.894:  98%|█████████▊| 49/50 [02:30<00:02,  2.93s/it]

[I 2026-08-22 01:11:20,924] Trial 48 finished with value: 337.5070138160095 and parameters: {'n_estimators': 700, 'learning_rate': 0.024757046521097677, 'max_depth': 4, 'min_child_weight': 8, 'subsample': 0.7518054694444615, 'colsample_bytree': 0.9843712832079435, 'gamma': 0.09237322253375163, 'reg_alpha': 1.044224972426472e-08, 'reg_lambda': 3.813838508883023}. Best is trial 39 with value: 326.8935821678058.


Best trial: 39. Best value: 326.894: 100%|██████████| 50/50 [02:32<00:00,  3.05s/it]

[I 2026-08-22 01:11:22,713] Trial 49 finished with value: 335.06959398384686 and parameters: {'n_estimators': 600, 'learning_rate': 0.017510501000403963, 'max_depth': 5, 'min_child_weight': 11, 'subsample': 0.7362621955406915, 'colsample_bytree': 0.9287011594084164, 'gamma': 0.17569314564422944, 'reg_alpha': 2.5469309101377956e-08, 'reg_lambda': 2.8844297443747817}. Best is trial 39 with value: 326.8935821678058.

BEST XGBOOST PARAMETERS
Best RMSE:
326.8935821678058

Best Parameters:
n_estimators: 800
learning_rate: 0.026340059352236846
max_depth: 5
min_child_weight: 11
subsample: 0.7263709147024497
colsample_bytree: 0.9493553396095371
gamma: 0.03818554358601167
reg_alpha: 4.0561758823329404e-08
reg_lambda: 2.310596649947486


In [24]:
# ============================================================
# Gradient Boosting + Optuna Hyperparameter Tuning
# ============================================================

import optuna
import numpy as np

from sklearn.ensemble import GradientBoostingRegressor
from sklearn.metrics import mean_squared_error


# ============================================================
# 1. Create time-based validation set
# ============================================================
# IMPORTANT:
# We only use X_tr_sc / y_tr for tuning.
# X_te_sc / y_te remains completely untouched.

val_size = int(len(X_tr_sc) * 0.20)

X_train_cv = X_tr_sc[:-val_size]
X_val_cv   = X_tr_sc[-val_size:]

y_train_cv = y_tr[:-val_size]
y_val_cv   = y_tr[-val_size:]


# ============================================================
# 2. Optuna objective function
# ============================================================

def objective(trial):

    # --------------------------------------------------------
    # Optuna will automatically search these parameters
    # --------------------------------------------------------

    params = {
        "n_estimators": trial.suggest_int(
            "n_estimators",
            100,
            800,
            step=50
        ),

        "learning_rate": trial.suggest_float(
            "learning_rate",
            0.01,
            0.20,
            log=True
        ),

        "max_depth": trial.suggest_int(
            "max_depth",
            2,
            8
        ),

        "min_samples_split": trial.suggest_int(
            "min_samples_split",
            2,
            30
        ),

        "min_samples_leaf": trial.suggest_int(
            "min_samples_leaf",
            1,
            20
        ),

        "subsample": trial.suggest_float(
            "subsample",
            0.6,
            1.0
        ),

        "max_features": trial.suggest_categorical(
            "max_features",
            [None, "sqrt", "log2"]
        ),

        "loss": trial.suggest_categorical(
            "loss",
            ["squared_error", "huber"]
        )
    }


    # --------------------------------------------------------
    # Create Gradient Boosting model
    # --------------------------------------------------------

    model = GradientBoostingRegressor(
        **params,
        random_state=42
    )


    # --------------------------------------------------------
    # Train on earlier observations
    # --------------------------------------------------------

    model.fit(
        X_train_cv,
        y_train_cv
    )


    # --------------------------------------------------------
    # Predict validation period
    # --------------------------------------------------------

    pred = model.predict(X_val_cv)


    # --------------------------------------------------------
    # RMSE = Optuna objective
    # Lower RMSE is better
    # --------------------------------------------------------

    rmse = np.sqrt(
        mean_squared_error(
            y_val_cv,
            pred
        )
    )

    return rmse


# ============================================================
# 3. Create Optuna study
# ============================================================

study_gb = optuna.create_study(
    direction="minimize",
    study_name="GradientBoosting_Electricity_Demand"
)


# ============================================================
# 4. Run optimization
# ============================================================

study_gb.optimize(
    objective,
    n_trials=50,
    show_progress_bar=True
)


# ============================================================
# 5. Show best result
# ============================================================

print("\n========================================")
print("BEST GRADIENT BOOSTING PARAMETERS")
print("========================================")

print("Best Validation RMSE:")
print(study_gb.best_value)

print("\nBest Parameters:")

for key, value in study_gb.best_params.items():
    print(f"{key}: {value}")

[I 2026-08-22 01:22:50,831] A new study created in memory with name: GradientBoosting_Electricity_Demand
  0%|          | 0/50 [00:00<?, ?it/s]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?
Best trial: 0. Best value: 755.259:   2%|▏         | 1/50 [00:08<07:16,  8.92s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 01:22:59,749] Trial 0 finished with value: 755.2585383758011 and parameters: {'n_estimators': 100, 'learning_rate': 0.020002719641915644, 'max_depth': 7, 'min_samples_split': 9, 'min_samples_leaf': 5, 'subsample': 0.605662163491419, 'max_features': 'sqrt', 'loss': 'huber'}. Best is trial 0 with value: 755.2585383758011.


Best trial: 1. Best value: 331.525:   4%|▍         | 2/50 [02:03<56:46, 70.96s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 01:24:54,139] Trial 1 finished with value: 331.5248896641915 and parameters: {'n_estimators': 700, 'learning_rate': 0.05680505060665195, 'max_depth': 3, 'min_samples_split': 18, 'min_samples_leaf': 20, 'subsample': 0.75476247111943, 'max_features': None, 'loss': 'squared_error'}. Best is trial 1 with value: 331.5248896641915.


Best trial: 2. Best value: 319.8:   6%|▌         | 3/50 [04:39<1:26:14, 110.09s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 01:27:30,795] Trial 2 finished with value: 319.7997244217298 and parameters: {'n_estimators': 350, 'learning_rate': 0.04859011499194396, 'max_depth': 8, 'min_samples_split': 13, 'min_samples_leaf': 9, 'subsample': 0.8043124138455698, 'max_features': None, 'loss': 'huber'}. Best is trial 2 with value: 319.7997244217298.


Best trial: 2. Best value: 319.8:   8%|▊         | 4/50 [05:50<1:12:25, 94.46s/it] d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 01:28:41,286] Trial 3 finished with value: 477.75522901440644 and parameters: {'n_estimators': 500, 'learning_rate': 0.015336527402830261, 'max_depth': 2, 'min_samples_split': 22, 'min_samples_leaf': 8, 'subsample': 0.9829562211473754, 'max_features': None, 'loss': 'huber'}. Best is trial 2 with value: 319.7997244217298.


Best trial: 2. Best value: 319.8:  10%|█         | 5/50 [06:15<52:11, 69.58s/it]  d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 01:29:06,769] Trial 4 finished with value: 374.0635929462189 and parameters: {'n_estimators': 250, 'learning_rate': 0.03869305101640386, 'max_depth': 7, 'min_samples_split': 22, 'min_samples_leaf': 19, 'subsample': 0.9670984241156386, 'max_features': 'sqrt', 'loss': 'huber'}. Best is trial 2 with value: 319.7997244217298.


Best trial: 2. Best value: 319.8:  12%|█▏        | 6/50 [06:25<36:03, 49.18s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 01:29:16,326] Trial 5 finished with value: 814.2378309245642 and parameters: {'n_estimators': 150, 'learning_rate': 0.014077322291485573, 'max_depth': 5, 'min_samples_split': 25, 'min_samples_leaf': 14, 'subsample': 0.8338472657482083, 'max_features': 'sqrt', 'loss': 'huber'}. Best is trial 2 with value: 319.7997244217298.


Best trial: 6. Best value: 309.922:  14%|█▍        | 7/50 [08:48<57:12, 79.82s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 01:31:39,232] Trial 6 finished with value: 309.9218432883534 and parameters: {'n_estimators': 700, 'learning_rate': 0.05128257080305278, 'max_depth': 4, 'min_samples_split': 20, 'min_samples_leaf': 17, 'subsample': 0.6972258345651443, 'max_features': None, 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  16%|█▌        | 8/50 [09:25<46:15, 66.08s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 01:32:15,912] Trial 7 finished with value: 336.86257999325977 and parameters: {'n_estimators': 150, 'learning_rate': 0.06974514910656142, 'max_depth': 4, 'min_samples_split': 17, 'min_samples_leaf': 8, 'subsample': 0.9005188085041842, 'max_features': None, 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  18%|█▊        | 9/50 [09:27<31:31, 46.13s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 01:32:18,156] Trial 8 finished with value: 1302.3743500484002 and parameters: {'n_estimators': 100, 'learning_rate': 0.020047452926571247, 'max_depth': 2, 'min_samples_split': 5, 'min_samples_leaf': 9, 'subsample': 0.8132505827426978, 'max_features': 'log2', 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  20%|██        | 10/50 [10:32<34:43, 52.08s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 01:33:23,562] Trial 9 finished with value: 327.00936070301634 and parameters: {'n_estimators': 250, 'learning_rate': 0.049979042841820524, 'max_depth': 4, 'min_samples_split': 9, 'min_samples_leaf': 18, 'subsample': 0.9620407198608829, 'max_features': None, 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  22%|██▏       | 11/50 [10:58<28:41, 44.14s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 01:33:49,692] Trial 10 finished with value: 469.0341685143506 and parameters: {'n_estimators': 750, 'learning_rate': 0.19487988608022933, 'max_depth': 6, 'min_samples_split': 30, 'min_samples_leaf': 1, 'subsample': 0.6038784249722582, 'max_features': 'log2', 'loss': 'squared_error'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  24%|██▍       | 12/50 [14:05<55:21, 87.42s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 01:36:56,098] Trial 11 finished with value: 354.5097426009685 and parameters: {'n_estimators': 500, 'learning_rate': 0.14396986345463336, 'max_depth': 8, 'min_samples_split': 14, 'min_samples_leaf': 14, 'subsample': 0.7146812660992586, 'max_features': None, 'loss': 'squared_error'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  26%|██▌       | 13/50 [17:33<1:16:31, 124.10s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 01:40:24,602] Trial 12 finished with value: 332.37321991633615 and parameters: {'n_estimators': 550, 'learning_rate': 0.0865339643147254, 'max_depth': 8, 'min_samples_split': 13, 'min_samples_leaf': 14, 'subsample': 0.6936211445901236, 'max_features': None, 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  28%|██▊       | 14/50 [19:07<1:08:58, 114.96s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 01:41:58,437] Trial 13 finished with value: 317.9676821917082 and parameters: {'n_estimators': 350, 'learning_rate': 0.03186761927525098, 'max_depth': 5, 'min_samples_split': 3, 'min_samples_leaf': 11, 'subsample': 0.7895600928215909, 'max_features': None, 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  30%|███       | 15/50 [21:32<1:12:20, 124.03s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 01:44:23,480] Trial 14 finished with value: 322.9904080391472 and parameters: {'n_estimators': 650, 'learning_rate': 0.03068089140998713, 'max_depth': 5, 'min_samples_split': 2, 'min_samples_leaf': 16, 'subsample': 0.6733784664211141, 'max_features': None, 'loss': 'squared_error'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  32%|███▏      | 16/50 [22:02<54:16, 95.77s/it]   d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 01:44:53,635] Trial 15 finished with value: 399.70079135815985 and parameters: {'n_estimators': 800, 'learning_rate': 0.025323381607337814, 'max_depth': 4, 'min_samples_split': 29, 'min_samples_leaf': 12, 'subsample': 0.762457825344688, 'max_features': 'log2', 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  34%|███▍      | 17/50 [24:03<56:52, 103.41s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 01:46:54,794] Trial 16 finished with value: 314.93320307998505 and parameters: {'n_estimators': 400, 'learning_rate': 0.09992193108619225, 'max_depth': 5, 'min_samples_split': 2, 'min_samples_leaf': 11, 'subsample': 0.8767671371885699, 'max_features': None, 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  36%|███▌      | 18/50 [25:52<55:55, 104.85s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 01:48:43,006] Trial 17 finished with value: 317.91910942343907 and parameters: {'n_estimators': 600, 'learning_rate': 0.1051493679920981, 'max_depth': 3, 'min_samples_split': 7, 'min_samples_leaf': 17, 'subsample': 0.8687356305261353, 'max_features': None, 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  38%|███▊      | 19/50 [26:16<41:36, 80.55s/it] d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 01:49:06,939] Trial 18 finished with value: 409.728907267554 and parameters: {'n_estimators': 400, 'learning_rate': 0.11314718233626665, 'max_depth': 6, 'min_samples_split': 24, 'min_samples_leaf': 6, 'subsample': 0.8936303865303855, 'max_features': 'sqrt', 'loss': 'squared_error'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  40%|████      | 20/50 [26:27<29:55, 59.85s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 01:49:18,566] Trial 19 finished with value: 427.218370550234 and parameters: {'n_estimators': 450, 'learning_rate': 0.07697321366918758, 'max_depth': 3, 'min_samples_split': 18, 'min_samples_leaf': 3, 'subsample': 0.6554903194392043, 'max_features': 'log2', 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  42%|████▏     | 21/50 [30:26<54:56, 113.67s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 01:53:17,719] Trial 20 finished with value: 323.6815064974448 and parameters: {'n_estimators': 650, 'learning_rate': 0.010711927122480884, 'max_depth': 6, 'min_samples_split': 10, 'min_samples_leaf': 12, 'subsample': 0.9233625361480367, 'max_features': None, 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  44%|████▍     | 22/50 [32:18<52:43, 112.97s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 01:55:09,049] Trial 21 finished with value: 323.69707192152913 and parameters: {'n_estimators': 600, 'learning_rate': 0.11105822841568307, 'max_depth': 3, 'min_samples_split': 6, 'min_samples_leaf': 17, 'subsample': 0.8888316098784832, 'max_features': None, 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  46%|████▌     | 23/50 [35:18<59:53, 133.09s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 01:58:09,055] Trial 22 finished with value: 313.80450747565664 and parameters: {'n_estimators': 750, 'learning_rate': 0.11105960036973325, 'max_depth': 4, 'min_samples_split': 6, 'min_samples_leaf': 17, 'subsample': 0.868579344982355, 'max_features': None, 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  48%|████▊     | 24/50 [38:26<1:04:48, 149.58s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 02:01:17,093] Trial 23 finished with value: 330.67720924406535 and parameters: {'n_estimators': 800, 'learning_rate': 0.1658642140586273, 'max_depth': 4, 'min_samples_split': 4, 'min_samples_leaf': 15, 'subsample': 0.8529682941372418, 'max_features': None, 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  50%|█████     | 25/50 [41:34<1:07:07, 161.11s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 02:04:25,121] Trial 24 finished with value: 312.9404135871333 and parameters: {'n_estimators': 700, 'learning_rate': 0.06552334661496795, 'max_depth': 5, 'min_samples_split': 2, 'min_samples_leaf': 19, 'subsample': 0.7612591039457968, 'max_features': None, 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  52%|█████▏    | 26/50 [43:56<1:02:07, 155.33s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 02:06:46,954] Trial 25 finished with value: 313.40069280701283 and parameters: {'n_estimators': 700, 'learning_rate': 0.06281472601058298, 'max_depth': 4, 'min_samples_split': 7, 'min_samples_leaf': 20, 'subsample': 0.7308639491160384, 'max_features': None, 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  54%|█████▍    | 27/50 [46:49<1:01:35, 160.67s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 02:09:40,082] Trial 26 finished with value: 328.5026224945255 and parameters: {'n_estimators': 700, 'learning_rate': 0.05869458157707553, 'max_depth': 5, 'min_samples_split': 11, 'min_samples_leaf': 20, 'subsample': 0.7411874468939557, 'max_features': None, 'loss': 'squared_error'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  56%|█████▌    | 28/50 [48:56<55:14, 150.64s/it]  d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 02:11:47,320] Trial 27 finished with value: 311.4032140378122 and parameters: {'n_estimators': 700, 'learning_rate': 0.040264192295685806, 'max_depth': 4, 'min_samples_split': 8, 'min_samples_leaf': 19, 'subsample': 0.6497537573786578, 'max_features': None, 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  58%|█████▊    | 29/50 [49:26<40:01, 114.37s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 02:12:17,075] Trial 28 finished with value: 379.4910537299187 and parameters: {'n_estimators': 600, 'learning_rate': 0.042850281552253476, 'max_depth': 6, 'min_samples_split': 16, 'min_samples_leaf': 18, 'subsample': 0.6415279460731661, 'max_features': 'log2', 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  60%|██████    | 30/50 [50:17<31:51, 95.55s/it] d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 02:13:08,715] Trial 29 finished with value: 361.5505484515417 and parameters: {'n_estimators': 750, 'learning_rate': 0.03474255801684362, 'max_depth': 7, 'min_samples_split': 20, 'min_samples_leaf': 18, 'subsample': 0.6385505119044269, 'max_features': 'sqrt', 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  62%|██████▏   | 31/50 [50:38<23:09, 73.14s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 02:13:29,546] Trial 30 finished with value: 431.14557898565994 and parameters: {'n_estimators': 650, 'learning_rate': 0.02851941755838676, 'max_depth': 3, 'min_samples_split': 27, 'min_samples_leaf': 16, 'subsample': 0.6999664794160431, 'max_features': 'sqrt', 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  64%|██████▍   | 32/50 [57:21<51:35, 171.98s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 02:20:12,159] Trial 31 finished with value: 311.9518442180803 and parameters: {'n_estimators': 700, 'learning_rate': 0.06603314998201933, 'max_depth': 4, 'min_samples_split': 8, 'min_samples_leaf': 20, 'subsample': 0.7270568788130866, 'max_features': None, 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  66%|██████▌   | 33/50 [1:02:05<58:14, 205.56s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 02:24:56,086] Trial 32 finished with value: 313.3371158382386 and parameters: {'n_estimators': 700, 'learning_rate': 0.04616472093087057, 'max_depth': 5, 'min_samples_split': 9, 'min_samples_leaf': 19, 'subsample': 0.7644851041952481, 'max_features': None, 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  68%|██████▊   | 34/50 [1:05:38<55:27, 207.96s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 02:28:29,645] Trial 33 finished with value: 310.0200577154203 and parameters: {'n_estimators': 800, 'learning_rate': 0.053779404983888064, 'max_depth': 4, 'min_samples_split': 15, 'min_samples_leaf': 19, 'subsample': 0.6645896555651332, 'max_features': None, 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  70%|███████   | 35/50 [1:09:12<52:24, 209.66s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 02:32:03,279] Trial 34 finished with value: 315.86551389021474 and parameters: {'n_estimators': 800, 'learning_rate': 0.02324095425319246, 'max_depth': 4, 'min_samples_split': 15, 'min_samples_leaf': 20, 'subsample': 0.6277960076085979, 'max_features': None, 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  72%|███████▏  | 36/50 [1:11:03<42:02, 180.21s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 02:33:54,763] Trial 35 finished with value: 349.76859063474626 and parameters: {'n_estimators': 750, 'learning_rate': 0.05266813527223908, 'max_depth': 2, 'min_samples_split': 12, 'min_samples_leaf': 16, 'subsample': 0.6739074687820765, 'max_features': None, 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  74%|███████▍  | 37/50 [1:13:55<38:28, 177.57s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 02:36:46,165] Trial 36 finished with value: 340.55126287996177 and parameters: {'n_estimators': 800, 'learning_rate': 0.03914623749171382, 'max_depth': 3, 'min_samples_split': 21, 'min_samples_leaf': 19, 'subsample': 0.6727235327390497, 'max_features': None, 'loss': 'squared_error'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  76%|███████▌  | 38/50 [1:16:46<35:06, 175.51s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 02:39:36,868] Trial 37 finished with value: 315.24644058765693 and parameters: {'n_estimators': 550, 'learning_rate': 0.08185776728871617, 'max_depth': 4, 'min_samples_split': 20, 'min_samples_leaf': 18, 'subsample': 0.7092827392715844, 'max_features': None, 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  78%|███████▊  | 39/50 [1:17:24<24:36, 134.25s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 02:40:14,847] Trial 38 finished with value: 379.21859934061325 and parameters: {'n_estimators': 650, 'learning_rate': 0.03660001319604759, 'max_depth': 4, 'min_samples_split': 24, 'min_samples_leaf': 15, 'subsample': 0.6191830671186139, 'max_features': 'sqrt', 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  80%|████████  | 40/50 [1:19:15<21:15, 127.55s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 02:42:06,757] Trial 39 finished with value: 347.36228082492585 and parameters: {'n_estimators': 750, 'learning_rate': 0.05788049274687437, 'max_depth': 2, 'min_samples_split': 8, 'min_samples_leaf': 20, 'subsample': 0.6574539721456848, 'max_features': None, 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  82%|████████▏ | 41/50 [1:21:25<19:14, 128.25s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 02:44:16,649] Trial 40 finished with value: 332.2997444811045 and parameters: {'n_estimators': 550, 'learning_rate': 0.04408293223225463, 'max_depth': 3, 'min_samples_split': 16, 'min_samples_leaf': 17, 'subsample': 0.7289851549978023, 'max_features': None, 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  84%|████████▍ | 42/50 [1:26:11<23:24, 175.58s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 02:49:02,654] Trial 41 finished with value: 316.97178172548996 and parameters: {'n_estimators': 700, 'learning_rate': 0.07024792971242866, 'max_depth': 5, 'min_samples_split': 18, 'min_samples_leaf': 19, 'subsample': 0.7879712593441455, 'max_features': None, 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  86%|████████▌ | 43/50 [1:30:02<22:24, 192.09s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 02:52:53,282] Trial 42 finished with value: 313.7745285403387 and parameters: {'n_estimators': 700, 'learning_rate': 0.06940682645218846, 'max_depth': 5, 'min_samples_split': 4, 'min_samples_leaf': 19, 'subsample': 0.6895420815709784, 'max_features': None, 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  88%|████████▊ | 44/50 [1:34:10<20:53, 208.89s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 02:57:01,354] Trial 43 finished with value: 313.8264123939503 and parameters: {'n_estimators': 750, 'learning_rate': 0.05227427757964801, 'max_depth': 4, 'min_samples_split': 11, 'min_samples_leaf': 18, 'subsample': 0.7452277566811785, 'max_features': None, 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  90%|█████████ | 45/50 [1:37:50<17:41, 212.32s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 03:00:41,703] Trial 44 finished with value: 311.9523706505087 and parameters: {'n_estimators': 650, 'learning_rate': 0.06189681192222465, 'max_depth': 4, 'min_samples_split': 14, 'min_samples_leaf': 20, 'subsample': 0.7757007091481648, 'max_features': None, 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  92%|█████████▏| 46/50 [1:40:51<13:31, 202.93s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 03:03:42,716] Trial 45 finished with value: 311.7840606173419 and parameters: {'n_estimators': 600, 'learning_rate': 0.04971613023738771, 'max_depth': 4, 'min_samples_split': 14, 'min_samples_leaf': 20, 'subsample': 0.8163675422968177, 'max_features': None, 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  94%|█████████▍| 47/50 [1:41:14<07:26, 148.97s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 03:04:05,769] Trial 46 finished with value: 430.8112366075696 and parameters: {'n_estimators': 600, 'learning_rate': 0.04621425220654144, 'max_depth': 3, 'min_samples_split': 12, 'min_samples_leaf': 15, 'subsample': 0.8190535651528209, 'max_features': 'log2', 'loss': 'squared_error'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  96%|█████████▌| 48/50 [1:43:26<04:47, 143.83s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 03:06:17,597] Trial 47 finished with value: 316.4284853723778 and parameters: {'n_estimators': 500, 'learning_rate': 0.08971386666433546, 'max_depth': 4, 'min_samples_split': 14, 'min_samples_leaf': 18, 'subsample': 0.6045218989221692, 'max_features': None, 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922:  98%|█████████▊| 49/50 [1:46:26<02:34, 154.69s/it]d:\Users\Md Mahfuzur Rahman\Desktop\Research\venv\Lib\site-packages\sklearn\ensemble\_gb.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


[I 2026-08-22 03:09:17,622] Trial 48 finished with value: 326.6077308596387 and parameters: {'n_estimators': 800, 'learning_rate': 0.03885954555145841, 'max_depth': 3, 'min_samples_split': 19, 'min_samples_leaf': 17, 'subsample': 0.6842522601080382, 'max_features': None, 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.


Best trial: 6. Best value: 309.922: 100%|██████████| 50/50 [1:49:11<00:00, 131.03s/it]

[I 2026-08-22 03:12:02,491] Trial 49 finished with value: 313.5304970029526 and parameters: {'n_estimators': 550, 'learning_rate': 0.05261011922863688, 'max_depth': 4, 'min_samples_split': 23, 'min_samples_leaf': 19, 'subsample': 0.717044747025304, 'max_features': None, 'loss': 'huber'}. Best is trial 6 with value: 309.9218432883534.

BEST GRADIENT BOOSTING PARAMETERS
Best Validation RMSE:
309.9218432883534

Best Parameters:
n_estimators: 700
learning_rate: 0.05128257080305278
max_depth: 4
min_samples_split: 20
min_samples_leaf: 17
subsample: 0.6972258345651443
max_features: None
loss: huber


In [ ]:
# ============================================================
# 6. Train FINAL Gradient Boosting model
# ============================================================

best_gb_params = study_gb.best_params

m_gb = GradientBoostingRegressor(
    **best_gb_params,
    random_state=42
)


# Train on ALL training data
m_gb.fit(
    X_tr_sc,
    y_tr
)


# ============================================================
# 7. Predict on untouched test data
# ============================================================

pred_gb = m_gb.predict(X_te_sc)


# ============================================================
# 8. Evaluate using your existing metrics function
# ============================================================

results.append(
    metrics(
        "Gradient Boosting Optuna",
        y_te,
        pred_gb
    )
)


print("\nGradient Boosting testing completed.")